### ISSUE TITLE
Merge inherited interface methods according to variance

### ISSUE THREAD

**Issue by alrz** [dotnet/runtime#102121]
In the following code `public void M(object arg)` should be enough to implement the interface but neither the compiler nor the runtime accept such code and an explicit impl is required to bridge the two overloads. ([sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAGxFBcAuIBOBXGAfAAQDYACAfQFgAoAb2oF9rqA6AYwQgGdOSAHTKADcIaAKYkI2NAHsJMTlBLBRAM2mZVUUQgAmsKGhIAFTNIDmmCGGolbJUQA8xMHTwDaAZQCenMWGYASrhoUGCiALrevqL+APLAAFairGjUdFR2JGzYvtJgJLC+EDCs4oLSUDoknj5+gcGhEVF1QTAhYcwAwnm8KKKYHv2CUKWcXT19mADiojD9IqI6AIJoaALAUqIgIGwymAAUAJQkALwk+zaZdgAMAIwk19cPT4+XtodvJAD0XyQAsqI0AALaSuT7MMLA0F8ATCMQkIGVUTAHxQMwkIohVgkT6ZcqVEgAcgAPH9oDAAHwAEkJ51xV1svgEMDMbnCEkwZk49PeJAIAGYcRk7OkGbYfv9ASCqsozIUJIYAgA1JYPBwAJmuAFZrjzvr9ujpxAoAF7ifkADnO1w1RA+woZBC1JAteoIABYmA6rgBJAAyZEejxAJDmAHdpIkCvI0MVSiR8VVOttduojnr/YHHlqQzpsLwMwGg0QQ+HIwlo0USmUKlUatF/K12k1ajFmPEkikUyk0/axSRM0HQCRWBAEAhhphDIVY9WE7WRxxuAOPHlRD62v0VBBSgADW7E8vJNAU7Z/fYAQmufbFg7u11z+cLWdutxDuiZJAAREyv8/HrcJYjmOE5QFOlazvGiaLlwPA+quYQbmImDbnuB5Hikp4gOeV43gyd63MOH5rN+TLqn+3qZHemohqO46TtOMZxjWBLsLBK5rkhW47qI+7EkysBmFhOHXv+1zqjmJAaKk3r0Pq9guCQ0gqCQkLSsYpgWFY2wkmSsDUoSXqZAQACcGK8MkUBjjAVjiFOnAWawVkIDZYRCv2M7MfOBKppg5zHAK7kinqEoAlCMqiHK8gKiQyqqjamoAOyJSFBqgsaUBmiQiXWg4iV4VcToukZt5FkG746BAnLMLqlF2Hedy0SBEFedB9YtA0YSRK2cSJMe3Z7OmdW2A1QHSZ8ckSrMVTKapUrQiY5iWGAA3qF6k2/NNSkqWxy6LVp1hUCwu08LAyGoeI/BCAsEhSLIxQKBIwBMjuhjKGoGgqFouj6IY8GcZuKE8XxAC0AAqFJpJ8oXzWC3qmaGohhpwCDSIYEDPWsr0JmBaDYGOQUMp5c7QX8dLDVcF5gxyZh6gFgqfKKdgbQpM0qWp0L/YhgMXfuZ7rfJW2zSdHHc+dwO3F6bBLqdPM8TC13wpIMhyI9GMvSkSiqOomjaHoMAGAOABCXCiJ8oS8AgMSzGgPAi1z65y2hh59ZhUNUCzQs7TLxum1LItndx8ZXXC4jK/d8iKOrWOa+9OtfXrv0DgAIv0QiLObYCW9bbRwSbnCiHAepivbCGO+LzsYSeRcUyXPsO1xQPOwJLKQ7QXoMILinC/XqewhnR1UNL7Ehzd4eq4ocefd9+uG50nyOM4rjVM0bZNo03UNu2rsyZkFtW2Eucp2ngiLDX/a3vnhfFwypcAxXvHoTvFLnxfVx32LQeP/xayCW3TPil+GFaU3JvQQlhnwbAwAEAjBIF9GyCAERIhRAodE4YUZoxxlOfGiC9TEygguMmFxa52CrjTOmfIGbegAVcGG4UtZRR4CIWKKo1SagtK6CmEpDQZSyjlfYNp8pumdJwsSwYSC6CqmYZgksKYNTfHyW41wq4FSomVa4Y1AQTS7mzOa9DkzYRKrYBGDlLLWVslJW2jlnKuXEHgpiJMFy+X8pQwmJAaGZDoepWU8pmFxTYdcZKqUSA8IxJlcQ/DBGqLsEVURcj1EPgkZVaqtV+zyKauOFqjiCTtTXp1FsW8Oz9R2D2A40SRrqM0bvZmOjtp6PUgY3yAsprd29uxeeg8gA==))

```cs
interface ISomeInterface<in T> {
  void M(T arg);
}

interface IBase : ISomeInterface<object>;

interface IDerived : IBase, ISomeInterface<string> {
   // uncomment to resolve the error but this shouldn't be necessary
   // void ISomeInterface<string>.M(string obj) => M(obj as object); // calls base
}

class C : IDerived {
  // error: method `M(string)` is not implemented
  public void M(object arg) { }
}
```

**Comment by CyrusNajmabadi** [comment]
i thnik this would take a language change.  i don't think the variance part of the lang spec supports this.    The variance part affects things like conversions, but not interface impls.

**Comment by alrz** [comment]
Indeed, but since identical method signatures from different interfaces are already merged, I think this would be just a relaxation of that rule. On the runtime, the variance already guaranteed that the call is safe.

**Comment by CyrusNajmabadi** [comment]
I'm not opposed to this being allowed. But would need runtime to push for wanting this. Otherwise, it's unlikely we would do it on the lang side.
